### PR TITLE
Allow possibility to add version to schema metadata

### PR DESCRIPTION
**Pull request by geoffreyaldebert**
This PR is referencing to https://github.com/opendatateam/udata/issues/2599 

It allows possibility to add version to schema metadata. With this implementation it is now possibile to differenciate two versions of a specific schema.

- we keep the official list of schema from schema.data.gouv.fr catalog
- we allow users to add version at the end of string like so : `etalab/schema-irve:1.0.3` or `etalab/schema-irve:2.0.0`

### CODE REVIEW

**Comment by geoffreyaldebert** [comment]
What needs to be done is allow possibility to select a specific version from User Interface (not yet implemented) on this PR. For that, schemas catalog may need to evolve by including a list of versions from specific schema.

**Comment by geoffreyaldebert** [comment]
PR en cours concernant les catalogues de schémas : https://framagit.org/opendataschema/catalog/-/merge_requests/4

**Comment by abulte** [comment]
I would go toward a pip syntax https://matiascodesal.com/blog/how-use-git-repository-pip-dependency/ (ie `@v1.0.0`)

**Comment by abulte** [comment]
I have to think some more about this, but we probably need a nested field with (at least):
- schema name
- schema version

This should ideally let use combine facets and list resources confirming to a schema and/or to a version.

**Comment by abulte** [comment]
We could introduce "schema catalog" in this field while we're at it, even if we set it to "schema.data.gouv.fr" by default and don't expose it in the UI.

**Comment by geoffreyaldebert** [comment]
Proposition taken into account @abulte 

Based on [ongoing PR on schema catalog repo](https://framagit.org/opendataschema/catalog/-/merge_requests/4) I have a waiting [PR on schema.data.gouv.fr repo](https://github.com/etalab/schema.data.gouv.fr/pull/154) to adapt our schemas catalog with new format. Example of this new type of schema catalog could be found [here](https://raw.githubusercontent.com/geoffreyaldebert/schema-test/master/schema-versions.json).

schema property is now a nested object containing two sub-properties `name` and `version` (we could add other properties like `schema_catalog` as suggested later with same structure.

What's this PR is doing :
- api is now working with this new structure and integrate version information (which is not mandatory)
- facet is now adapted with this new structure and we can now filter by `schema` or `schema_version`

What I would like to have advice on : 
- As we change format of `schema` field. Is this a problem for existing data with `schema` property already filled ? Probably.. how can we get out of it ?

What needs to be done : 
- [ ] Modify modal in https://github.com/etalab/udata-schema-gouvfr to be adapted with this new structure
- [ ] Adapt form when creating a new resource with select button for schema version which is linked to schema name select field. (we may not have to do it if we decide that all entries for schema data should pass by new service publier.data.gouv.fr)

File: udata/core/dataset/forms.py [hunk 1]
```diff
@@ -74,8 +81,8 @@ class BaseResourceForm(ModelForm):
         _('Publication date'),
         description=_('The publication date of the resource'))
     extras = fields.ExtrasField()
-    schema = fields.StringField(
-        _('Schema'),
+    schema = fields.DictField(
+        _('Schema'),
```
Comment (abulte) [inline udata/core/dataset/forms.py:85]:
```suggestion
        _('Schema'),
```

File: udata/core/dataset/search.py [hunk 2]
```diff
@@ -85,6 +85,7 @@ class Meta:
         'description': String(),
         'format': String(index='not_analyzed'),
         'schema': String(index='not_analyzed'),
+        'schema_version': String(index='not_analyzed')
```
Comment (abulte) [inline udata/core/dataset/search.py:88]:
```suggestion
        'schema_version': String(index='not_analyzed'),
```

File: udata/core/dataset/search.py [hunk 3]
```diff
@@ -189,7 +191,7 @@ def serialize(cls, dataset):
             image_url = owner.avatar(40, external=True)
 
         certified = organization and organization.certified
-
+
```
Comment (abulte) [inline udata/core/dataset/search.py:194]:
```suggestion

```

File: udata/core/dataset/search.py [hunk 4]
```diff
@@ -203,7 +205,8 @@ def serialize(cls, dataset):
                     'description': r.description,
                     'format': r.format,
                     'type': r.type,
-                    'schema': r.schema,
+                    'schema': dict(r.schema).get('name',{}),
+                    'schema_version': dict(r.schema).get('version',{})
```
Comment (abulte) [inline udata/core/dataset/search.py:209]:
```suggestion
                    'schema_version': dict(r.schema).get('version',{}),
```

File: udata/core/dataset/models.py [hunk 5]
```diff
@@ -799,7 +799,7 @@ def objects():
         else:
             schemas = response.json().get('schemas', [])
             content = [
-                {'id': s['name'], 'label': s['title']} for s in schemas
+                {'id': s['name'], 'label': s['title'], 'versions': [d['version_name'] for d in s['versions']]} for s in schemas
```
Comment (abulte) [inline udata/core/dataset/models.py:802]:
```suggestion
                {
                    'id': s['name'], 
                    'label': s['title'], 
                    'versions': [d['version_name'] for d in s['versions']],
                } for s in schemas
```

File: udata/core/dataset/api.py [hunk 6]
```diff
@@ -44,7 +44,7 @@
     resource_fields,
     resource_type_fields,
     upload_fields,
-    schema_fields,
+    schema_fields
```
Comment (abulte) [inline udata/core/dataset/api.py:47]:
```suggestion
    schema_fields,
```

File: udata/core/dataset/api_fields.py [hunk 7]
```diff
@@ -214,7 +215,9 @@
     'label': fields.String(description='The resource type display name')
 })
 
+
 schema_fields = api.model('Schema', {
     'id': fields.String(description='The schema identifier'),
-    'label': fields.String(description='The schema display name')
+    'label': fields.String(description='The schema display name'),
+    'versions': fields.List(fields.String, description='The available versions of the schema')
```
Comment (abulte) [inline udata/core/dataset/api_fields.py:222]:
```suggestion
    'versions': fields.List(fields.String, description='The available versions of the schema'),
```

**Review by abulte — COMMENTED** [review]
Code style nitpicks

**Review by quaxsze — APPROVED** [review]

File: udata/core/dataset/forms.py [hunk 8]
```diff
@@ -12,7 +12,7 @@
     Dataset, Resource, License, Checksum, CommunityResource,
     UPDATE_FREQUENCIES, DEFAULT_FREQUENCY, RESOURCE_FILETYPES, CHECKSUM_TYPES,
     LEGACY_FREQUENCIES, RESOURCE_TYPES, TITLE_SIZE_LIMIT, DESCRIPTION_SIZE_LIMIT,
-    ResourceSchema,
+    ResourceSchema
```
Comment (abulte) [inline udata/core/dataset/forms.py:15]:
```suggestion
    ResourceSchema,
```

File: udata/core/dataset/forms.py [hunk 9]
```diff
@@ -30,17 +30,35 @@ def normalize_format(data):
...
-        message = _('Schema "{schema}" is not an allowed value. Allowed values: {values}')
-        raise validators.ValidationError(message.format(
-            schema=schema,
-            values=', '.join(allowed_schemas)
-        ))
-
+    if schema != {}:
+        allowed_schemas = [s['id'] for s in ResourceSchema.objects()]
+        if schema.get('name') not in allowed_schemas:
+            print(schema)
```
Comment (abulte) [inline udata/core/dataset/forms.py:38]:
```suggestion
```

File: udata/core/dataset/forms.py [hunk 10]
```diff
@@ -30,17 +30,35 @@ def normalize_format(data):
...
+            message = _('Schema name "{schema}" is not an allowed value. Allowed values: {values}')
+            raise validators.ValidationError(message.format(
+                schema=schema.get('name'),
+                values=', '.join(allowed_schemas)
+            ))
+
+        allowed_versions = [d['versions'] for d in ResourceSchema.objects() if d['id'] == schema.get('name')][0]
```
Comment (abulte) [inline udata/core/dataset/forms.py:45]: I think you need to protect against this list being empty.

File: udata/core/dataset/forms.py [hunk 11]
```diff
@@ -30,17 +30,35 @@ def normalize_format(data):
...
+        allowed_versions.append('latest')
+        if "version" in schema:
```
Comment (abulte) [inline udata/core/dataset/forms.py:47]:
```suggestion
        if 'version' in schema:
```

File: udata/core/dataset/forms.py [hunk 12]
```diff
@@ -30,17 +30,35 @@ def normalize_format(data):
...
+            if schema.get('version') not in allowed_versions:
+                message = _('Version "{version}" is not an allowed value. Allowed values: {values}')
+                raise validators.ValidationError(message.format(
+                    version=schema.get('version'),
+                    values=', '.join(allowed_versions)
+                ))
+        properties = ["name","version"]
```
Comment (abulte) [inline udata/core/dataset/forms.py:54]:
```suggestion
        properties = ['name', 'version']
```

File: udata/core/dataset/forms.py [hunk 13]
```diff
@@ -30,17 +30,35 @@ def normalize_format(data):
...
+        for prop in schema:
+            if(prop not in properties):
```
Comment (abulte) [inline udata/core/dataset/forms.py:56]:
```suggestion
            if prop not in properties:
```

File: udata/core/dataset/forms.py [hunk 14]
```diff
@@ -30,17 +30,35 @@ def normalize_format(data):
...
+                message = _('Sub-property "{prop}" is not allowed value in schema field. Allowed values is : {properties}')
+                raise validators.ValidationError(message.format(
+                    prop=prop,
+                    properties=', '.join(properties)
```
Comment (abulte) [inline udata/core/dataset/forms.py:60]:
```suggestion
                    properties=', '.join(properties),
```

File: udata/core/dataset/search.py [hunk 15]
```diff
@@ -203,7 +204,8 @@ def serialize(cls, dataset):
                     'description': r.description,
                     'format': r.format,
                     'type': r.type,
-                    'schema': r.schema,
+                    'schema': dict(r.schema).get('name',{}),
```
Comment (abulte) [inline udata/core/dataset/search.py:207]:
```suggestion
                    'schema': dict(r.schema).get('name', {}),
```

File: udata/core/dataset/search.py [hunk 16]
```diff
@@ -203,7 +204,8 @@ def serialize(cls, dataset):
...
+                    'schema_version': dict(r.schema).get('version',{}),
```
Comment (abulte) [inline udata/core/dataset/search.py:208]:
```suggestion
                    'schema_version': dict(r.schema).get('version', {}),
```

File: udata/migrations/2021-04-08-update-schema-with-new-structure.py [hunk 17]
```diff
@@ -0,0 +1,32 @@
+'''
+The purpose here is to fill every resource with a fs_filename string field.
```
Comment (abulte) [inline udata/migrations/2021-04-08-update-schema-with-new-structure.py:2]: Update this.

File: udata/migrations/2021-04-08-update-schema-with-new-structure.py [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+def migrate(db):
+    log.info('Processing resources.')
+
+    datasets = Dataset.objects().no_cache().timeout(False)
+    for dataset in datasets:
+        save_res = False
+        for resource in dataset.resources:
+            if hasattr(resource, 'schema'):
+                schema = resource.schema
+                resource.schema = { "name": None }
```
Comment (abulte) [inline udata/migrations/2021-04-08-update-schema-with-new-structure.py:20]:
```suggestion
                resource.schema = {'name': None}
```

File: udata/migrations/2021-04-08-update-schema-with-new-structure.py [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+                if schema is not None: 
+                    if isinstance(schema, str):
+                        resource.schema = { "name": schema }
```
Comment (abulte) [inline udata/migrations/2021-04-08-update-schema-with-new-structure.py:23]:
```suggestion
                        resource.schema = {'name': schema}   
```

File: udata/migrations/2021-04-08-update-schema-with-new-structure.py [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+
+    datasets = Dataset.objects().no_cache().timeout(False)
+    for dataset in datasets:
+        save_res = False
+        for resource in dataset.resources:
+            if hasattr(resource, 'schema'):
+                schema = resource.schema
+                resource.schema = { "name": None }
+                if schema is not None: 
+                    if isinstance(schema, str):
```
Comment (abulte) [inline udata/migrations/2021-04-08-update-schema-with-new-structure.py:22]:
```suggestion
                if schema is not None and isinstance(schema, str):
```

File: udata/tests/api/test_datasets_api.py [hunk 21]
```diff
@@ -1473,13 +1473,14 @@ def test_dataset_schemas_api_list(self, api, rmock, app):
         app.config['SCHEMA_CATALOG_URL'] = 'https://example.com/schemas'
 
         rmock.get('https://example.com/schemas', json={
-            'schemas': [{"name": "etalab/schema-irve", "title": "Schéma IRVE"}]
+            'schemas': [{"name": "etalab/schema-irve", "title": "Schéma IRVE","versions":[{"version_name": "1.0.0"},{"version_name": "1.0.1"},{"version_name": "1.0.2"}]}]
         })
-
+        print("jjj")
```
Comment (abulte) [inline udata/tests/api/test_datasets_api.py:1478]:
```suggestion
```

File: udata/tests/api/test_datasets_api.py [hunk 22]
```diff
@@ -1503,15 +1504,16 @@ def test_dataset_schemas_api_list_error_no_cache(self, api, rmock):
...
         cache_mock_set = mocker.patch.object(cache, 'set')
-        mocker.patch.object(cache, 'get', return_value=[{"id": "etalab/schema-irve", "label": "Schéma IRVE"}])
+        mocker.patch.object(cache, 'get', return_value=[{"id": "etalab/schema-irve", "label": "Schéma IRVE","versions":["1.0.0","1.0.1","1.0.2"]}])
 
         # Fill cache
         rmock.get('https://example.com/schemas', json={
-            'schemas': [{"name": "etalab/schema-irve", "title": "Schéma IRVE"}]
+            'schemas': [{"name": "etalab/schema-irve", "title": "Schéma IRVE","versions":[{"version_name": "1.0.0"},{"version_name": "1.0.1"},{"version_name": "1.0.2"}]}]
         })
+        print('yolo')
```
Comment (abulte) [inline udata/tests/api/test_datasets_api.py:1513]:
```suggestion
```

File: udata/tests/api/test_datasets_api.py [hunk 23]
```diff
@@ -1473,13 +1473,14 @@ def test_dataset_schemas_api_list(self, api, rmock, app):
         app.config['SCHEMA_CATALOG_URL'] = 'https://example.com/schemas'
 
         rmock.get('https://example.com/schemas', json={
-            'schemas': [{"name": "etalab/schema-irve", "title": "Schéma IRVE"}]
+            'schemas': [{"name": "etalab/schema-irve", "title": "Schéma IRVE","versions":[{"version_name": "1.0.0"},{"version_name": "1.0.1"},{"version_name": "1.0.2"}]}]
```
Comment (abulte) [inline udata/tests/api/test_datasets_api.py:1476]: review indentation and spacing and quotes (which were already wrong) here

File: udata/tests/api/test_datasets_api.py [hunk 24]
```diff
@@ -1473,13 +1473,14 @@ def test_dataset_schemas_api_list(self, api, rmock, app):
...
         })
-
+        print("jjj")
         response = api.get(url_for('api.schemas'))
 
+        print(response.json)
         assert200(response)
-        assert response.json == [{"id": "etalab/schema-irve", "label": "Schéma IRVE"}]
+        assert response.json == [{"id": "etalab/schema-irve", "label": "Schéma IRVE","versions":["1.0.0","1.0.1","1.0.2"]}]
```
Comment (abulte) [inline udata/tests/api/test_datasets_api.py:1483]: review indentation and spacing and quotes (which were already wrong) here

File: udata/tests/api/test_datasets_api.py [hunk 25]
```diff
@@ -1503,15 +1504,16 @@ def test_dataset_schemas_api_list_error_no_cache(self, api, rmock):
     @pytest.mark.options(SCHEMA_CATALOG_URL='https://example.com/schemas')
     def test_dataset_schemas_api_list_error_w_cache(self, api, rmock, mocker):
         cache_mock_set = mocker.patch.object(cache, 'set')
-        mocker.patch.object(cache, 'get', return_value=[{"id": "etalab/schema-irve", "label": "Schéma IRVE"}])
+        mocker.patch.object(cache, 'get', return_value=[{"id": "etalab/schema-irve", "label": "Schéma IRVE","versions":["1.0.0","1.0.1","1.0.2"]}])
 
         # Fill cache
         rmock.get('https://example.com/schemas', json={
-            'schemas': [{"name": "etalab/schema-irve", "title": "Schéma IRVE"}]
+            'schemas': [{"name": "etalab/schema-irve", "title": "Schéma IRVE","versions":[{"version_name": "1.0.0"},{"version_name": "1.0.1"},{"version_name": "1.0.2"}]}]
```
Comment (abulte) [inline udata/tests/api/test_datasets_api.py:1511]: review indentation and spacing and quotes (which were already wrong) here

**Review by abulte — COMMENTED** [review]
💪 😎 💹

**Pull request merged**